### PR TITLE
Refactor to separate provider and connector classes

### PR DESCRIPTION
**Pull request by nenadV91**
# Summary

Refactor to separate Provider and Connector classes for ledger.

### CODE REVIEW

File: src/cow-react/modules/wallet/web3-react/connectors/Ledger/provider.tsx [hunk 1]
```diff
@@ -26,63 +40,14 @@ export function generateDerivationPaths(context: { accountsOffset: number; accou
...
-      console.error(error)
-    })
+  constructor(options: LedgerOptions) {
+    this.options = options
   }
-}
 
-export class Ledger extends Connector {
-  static getModule(): Promise<LedgerModules> {
+  getModule = (): Promise<LedgerModules> => {
```
Comment (shoom3301) [inline src/cow-react/modules/wallet/web3-react/connectors/Ledger/provider.tsx:50]: Actually, would be better to move it to a function

File: src/cow-react/modules/wallet/web3-react/connectors/Ledger/connector.tsx [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      method: 'net_version',
+    })) as string
+
+    return +chainId
+  }
+
+  async activate() {
+    if (this.isActivated) return
+
+    const provider = new LedgerProvider(this.options)
```
Comment (shoom3301) [inline src/cow-react/modules/wallet/web3-react/connectors/Ledger/connector.tsx:68]: Please, get rid of all `any`

File: src/cow-react/modules/wallet/web3-react/connectors/Ledger/provider.tsx [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      // @ts-ignore
+      import('@ledgerhq/hw-transport-u2f'),
+      // @ts-ignore
+      import('@ledgerhq/hw-transport-webhid'),
+      // @ts-ignore
+      import('@ledgerhq/web3-subprovider'),
+      // @ts-ignore
+      import('@ledgerhq/hw-app-eth'),
+    ])
+      // eslint-disable-next-line @typescript-eslint/naming-convention
```
Comment (shoom3301) [inline src/cow-react/modules/wallet/web3-react/connectors/Ledger/provider.tsx:55]: Would be great to get rid off this and `@ts-ignore`. If it's possible

File: src/cow-react/modules/wallet/web3-react/connectors/Ledger/provider.tsx [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+export interface EcdsaSignature {
+  r: string
+  s: string
+  v: number
+}
+
+// TODO: use current chain Id
+const currentChainId = 1
```
Comment (shoom3301) [inline src/cow-react/modules/wallet/web3-react/connectors/Ledger/provider.tsx:25]: Please, don't forget about this

File: src/cow-react/modules/wallet/web3-react/connectors/Ledger/provider.tsx [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    _domain: TypedDataDomain,
+    types: Record<string, Array<TypedDataField>>,
+    value: Record<string, any>,
+    primaryType: string
+  ): Promise<string> => {
+    const module = await _getModule()
+    const transport = await module.transportFactory.create()
+
+    // TODO: https://etherscan.io/address/0x9008d19f58aabd9ed0d60971565aa8510560ab41#readContract
+    const domainSeparator = '0xc078f884a2676e1345748b1feace7b0abee5d00ecadb6e574dcdd109a63e8943'
```
[the remark this answers is not in the document]
Comment (nenadV91) [reply src/cow-react/modules/wallet/web3-react/connectors/Ledger/provider.tsx]: Updated this to get the value from the contract

File: src/cow-react/modules/wallet/web3-react/connectors/Ledger/provider.tsx (hunk 4 above)
Comment (nenadV91) [reply src/cow-react/modules/wallet/web3-react/connectors/Ledger/provider.tsx]: Changed it to pass it from the connector constructor or use default.

File: src/cow-react/modules/wallet/web3-react/connectors/Ledger/provider.tsx (hunk 3 above)
Comment (nenadV91) [reply src/cow-react/modules/wallet/web3-react/connectors/Ledger/provider.tsx:58]: Didn't find types for few of those 😞

File: src/cow-react/modules/wallet/web3-react/connectors/Ledger/provider.tsx [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    this.rpcUrl = this.options.rpc?.[this.options.chainId || 1] || DEFAULT_RPC
+    this.chainId = this.options.chainId || DEFAULT_CHAIN_ID
+  }
+
+  async getDomainSeparator() {
+    const provider = new JsonRpcProvider(this.rpcUrl)
+    const address = GP_SETTLEMENT_CONTRACT_ADDRESS[this.chainId] || ''
+    const contract = getContract(address, GPv2_SETTLEMENT_ABI, provider)
+
+    return contract.domainSeparator()
```
Comment (anxolin) [inline src/cow-react/modules/wallet/web3-react/connectors/Ledger/provider.tsx:86]: Wouldn't this just work and be easier and faster?

```
import { domain, TypedDataDomain } from '@cowprotocol/contracts'
```

File: src/cow-react/modules/wallet/web3-react/connectors/Ledger/connector.tsx [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+      forceCreate: false,
+    }
+  ): Promise<Provider> {
+    if (!this.provider) {
+      if (forceCreate) {
+        this.activate()
+      }
+
+      return new Promise((resolve) => {
+        this.providerRequests.push(resolve)
```
Comment (anxolin) [inline src/cow-react/modules/wallet/web3-react/connectors/Ledger/connector.tsx:42]: what about the rejection?

**Review by anxolin — COMMENTED** [review]
Hei Nenad, 

I guess you also tried using the iframe provider, right? 
https://www.npmjs.com/package/@ledgerhq/iframe-provider

I quickly skimmed in their docs and they point to actually a web3-react connect Lido did. Would that work for us? Cause if it works, implementations seems way easier than the one suggested here tbh. I haven't digged deeper into this, so just curious on this experiments you are doing. 

I also thought Lido have some code that might be worth looking at, because they have lots of connectors, so I think we can really copy a lot of their logic. Isn't it? 

https://github.com/lidofinance/lido-js-sdk/blob/6e599d0dddccf145287174544808e707c5e54dc5/packages/web3-react/src/context/connectors.tsx

 
image

**Comment by nenadV91** [comment]
Closing as its not relevant enymore

**Pull request closed without merging**